calrissian/executor.py

```diff
@@ -1,40 +1,416 @@
+        :return: True if the queue is empty, False otherwise
+        """
+        return len(self.jobs) == 0
 
-    def run_jobs(self, *args, **kwargs):
-        self.report_usage('run_jobs')
-        return super(CalrissianExecutor, self).run_jobs(*args, **kwargs)
+    def sorted_jobs(self):
+        """
+        Produces a list of the jobs in the queue, ordered by self.priority (RAM or CORES) and ascending or descending
+        :return:
+        """
+        return sorted(self.jobs.items(), key=lambda item: getattr(item[1], self.priority), reverse=self.descending)
 
-    def execute(self, *args, **kwargs):
-        self.report_usage('execute')
-        return super(CalrissianExecutor, self).execute(*args, **kwargs)
+    def dequeue(self, resource_limit):
+        """
+        Collects jobs from the sorted list that fit together within the specified resource limit.
+        Removes (pop) collected jobs from the queue (pop).
+        May return an empty dictionary if queue is empty or no jobs fit.imit
+        :param resource_limit: A Resource object
+        :return: Dictionary of {Job:Resources}
+        """
+        jobs = {}
+        for job, resource in self.sorted_jobs():
+            if resource_limit - resource >= Resources.EMPTY:
+                jobs[job] = resource
+                resource_limit = resource_limit - resource
+        for job in jobs:
+            self.jobs.pop(job)
+        return jobs
+
+
+class ThreadPoolJobExecutor(JobExecutor):
+    """
+    A cwltool JobExecutor subclass that uses concurrent.futures.ThreadPoolExecutor
+
+    concurrent.futures was introduced in Python 3.2 and provides high-level interfaces (Executor, Future) for launching
+    and managing asynchronous and parallel tasks. The ThreadPoolExecutor maintains a pool of reusable threads to execute
+    tasks, reducing the overall number of threads that are created in the cwltool process.
+
+    Relevant: https://github.com/common-workflow-language/cwltool/issues/888
+    """
+
+    def __init__(self, total_ram, total_cores, max_workers=None):
+        """
+        Initialize a ThreadPoolJobExecutor
+        :param total_ram: RAM limit in megabytes for concurrent jobs
+        :param total_cores: cpu core count limit for concurrent jobs
+        :param max_workers: Number of worker threads to create. Set to None to use Python's default of 5xcpu count, which
+        should be sufficient. Setting max_workers too low can cause deadlocks.
+
+        See https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.ThreadPoolExecutor
+        """
+        super(ThreadPoolJobExecutor, self).__init__()
+        self.max_workers = max_workers
+        self.jrq = JobResourceQueue()
+        self.exceptions = Queue()
+        self.total_resources = Resources(total_ram, total_cores)
+        self.available_resources = Resources(total_ram, total_cores) # start with entire pool available
+        self.resources_lock = threading.RLock()
```

Comment:
Are we making use of the reentrant functionality of this lock?

Reply:
No - originally I did write it to use the re-entrant functionality but the allocate/restore got reworked a few times and no longer needs re-entrant locks. Good catch.